openlibrary/accounts/model.py

```diff
@@ -267,3 +276,462 @@ def set_bot_flag(self, flag):
+
+        if len(lower_match):
+            return cls(lower_match[0])
+
+        return None
+
+    @classmethod
+    def get_by_link(cls, link, test=False):
+        ol_accounts = web.ctx.site.store.values(
+            type="account", name="internetarchive_itemname", value=link)
+        return cls(ol_accounts[0]) if ol_accounts else None
+
+    @classmethod
+    def get_by_email(cls, email, test=False):
+        """the email stored in account doc is case-sensitive.
+        The lowercase of email is used in the account-email document.
+        querying that first and taking the username from there to make
+        the email search case-insensitive.
+
+        There are accounts with case-variation of emails. To handle
+        those, searching with the original case and using lower case
+        if that fails.
+        """
+        email_doc = (web.ctx.site.store.get("account-email/" + email) or
+                     web.ctx.site.store.get("account-email/" + email.lower()))
+        if email_doc and 'username' in email_doc:
+            doc = web.ctx.site.store.get("account/" + email_doc['username'])
+            return cls(doc) if doc else None
+        return None
+
+    @property
+    def verified(self):
+        return not (getattr(self, 'status', '') == 'pending')
+
+    @property
+    def blocked(self):
+        return getattr(self, 'status', '') == 'blocked'
+
+    def unlink(self):
+        """Careful, this will save any other changes to the ol user object as
+        well
+        """
+        _ol_account = web.ctx.site.store.get(self._key)
+        _ol_account['internetarchive_itemname'] = None
+        web.ctx.site.store[self._key] = _ol_account
+        self.internetarchive_itemname = None
+
+    def link(self, itemname):
+        """Careful, this will save any other changes to the ol user object as
+        well
+        """
+        _ol_account = web.ctx.site.store.get(self._key)
+        _ol_account['internetarchive_itemname'] = itemname
+        web.ctx.site.store[self._key] = _ol_account
+        self.internetarchive_itemname = itemname
+
+    @classmethod
+    def authenticate(cls, email, password, test=False):
+        ol_account = cls.get(email=email, test=test)
+        if not ol_account:
+            return "account_not_found"
+        if ol_account.is_blocked():
+            return "account_blocked"
+        try:
+            web.ctx.site.login(ol_account.username, password)
+        except ClientException, e:
+            code = e.get_data().get("code")
+            return code
+        else:
+            return "ok"
+
+class InternetArchiveAccount(web.storage):
+
+    def __init__(self, **kwargs):
+        for k in kwargs:
+            setattr(self, k, kwargs[k])
+
+    @classmethod
+    def create(cls, screenname, email, password, retries=0,
+               verified=False, test=None):
+        screenname = screenname.replace('@', '')  # remove IA @
```

Comment:
Good point. Updates to `screenname = screenname[1:] if screenname[0] == '@' else screenname`